### PR TITLE
feat(js): fetchTokenByRefreshToken

### PR DESCRIPTION
**Pull request by demonzoo**
## Summary

Implement js core function: fetchTokenByRefreshToken

## Linear Issue Reference

https://linear.app/silverhand/issue/LOG-771/fetch-tokens-by-refresh-token

## Testing

- [x] Passed all unit test cases

### CODE REVIEW

File: packages/js/jest.config.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ module.exports = {
   coverageReporters: ['text-summary', 'lcov'],
   moduleFileExtensions: ['ts', 'js'],
   setupFilesAfterEnv: ['./jest.setup.js', 'jest-matcher-specific-error'],
+  testEnvironment: 'jsdom',
```
Comment (gao-sun) [inline packages/js/jest.config.js:6]: @IceHe shall we align the usage of setting environment? I saw file level declarations as well.

File: packages/js/src/core/fetch-token.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  clientId,
+  tokenEndPoint,
+  refreshToken,
+  resource,
+  scope,
+}: FetchTokenByRefreshTokenParameters): Promise<RefreshTokenTokenResponse> => {
+  const parameters = new URLSearchParams();
+  parameters.append('client_id', clientId);
+  parameters.append('refresh_token', refreshToken);
+  if (resource) {
```
Comment (gao-sun) [inline packages/js/src/core/fetch-token.ts:41]: empty line before this line?

File: packages/js/src/core/fetch-token.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    parameters.append('resource', resource);
+  }
+
+  if (scope?.length) {
+    parameters.append('scope', scope.join(' '));
+  }
+
+  const requester = createRequester();
```
Comment (gao-sun) [inline packages/js/src/core/fetch-token.ts:49]: Will it be better to use a shared requester instead of creating every time?

File: packages/js/src/core/fetch-token.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    return {
+      accessToken: access_token,
+      refreshToken: refresh_token,
+      idToken: id_token,
+      scope: scope.split(' '),
+      expiresIn: expires_in,
+    };
+  } catch (error: unknown) {
+    if (error instanceof StructError) {
+      throw new LogtoError('struct_verification.response_struct_mismatched');
```
Comment (gao-sun) [inline packages/js/src/core/fetch-token.ts:70]: We can let superstruct throw the error directly instead of wrapping another layer to keep simplicity.

File: packages/playground/package.json [hunk 5]
```diff
@@ -10,7 +10,7 @@
     "dev:razzle": "razzle start",
     "start": "concurrently -c \"blue,cyan\" \"pnpm:dev:tsc\" -k \"pnpm:dev:razzle\"",
     "start:prod": "NODE_ENV=production node build/server.js",
-    "build": "tsc -b && razzle build",
+    "build": "tsc -b && razzle build --noninteractive",
```
Comment (gao-sun) [inline packages/playground/package.json:13]: This should be submit in a separate PR, also not sure if we need this. I'm ok with both.

Comment (demonzoo) [reply packages/playground/package.json:13]: I think we do need this, as sometimes the CLI prompts for yes/no. And if we don't disable this, the CI will fail, which was exactly the case for me last night.

image

File: packages/js/src/core/fetch-token.ts (hunk 4 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:70]: OK, fine for me. Will update shortly

File: packages/js/src/core/fetch-token.ts (hunk 2 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:41]: OK, let me take a look at the linting rules

File: packages/js/jest.config.js (hunk 1 above)
Comment (demonzoo) [reply packages/js/jest.config.js:6]: If I remember correctly, the file level declarations was just for the node environment. So declaring the default test environment as 'jsdom', and use file level declarations for node environment as an exception makes sense to me.

File: packages/js/src/core/fetch-token.ts (hunk 3 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:49]: Like a singleton instance? If that's necessary, I think it would be better to be done in the `createRequester` itself

File: packages/js/jest.config.js (hunk 1 above)
Comment (gao-sun) [reply packages/js/jest.config.js:6]: why do we need file level "node" declaration?

Comment (simeng-li) [reply packages/js/jest.config.js:6]: actually since js core SDK is not meant for dom only. we could leave it as default ?

File: packages/js/src/core/fetch-token.ts (hunk 3 above)
Comment (gao-sun) [reply packages/js/src/core/fetch-token.ts:49]: Imaging we'd like to use some lib "foo" to send requests, then how do we achieve it with current design?

File: packages/playground/package.json (hunk 5 above)
Comment (gao-sun) [reply packages/playground/package.json:13]: Interesting, why CI passed before?

File: packages/js/jest.config.js (hunk 1 above)
Comment (demonzoo) [reply packages/js/jest.config.js:6]: Please take a look at [this discussion in PR#137](https://github.com/logto-io/js/pull/137/files#r791602491)
Looks like we have a node env specific error handling

File: packages/js/src/core/fetch-token.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+  if (scope?.length) {
+    parameters.append('scope', scope.join(' '));
+  }
+
+  const requester = createRequester();
+  const response = await requester(tokenEndPoint, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/x-www-form-urlencoded',
```
Comment (simeng-li) [inline packages/js/src/core/fetch-token.ts:53]: Actually I am thinking, as for the `content_type` header, should it be placed to some common place or encapsulated within the requester itself by default?  So no need for us to provide it manually in every api.

File: packages/js/src/core/fetch-token.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    },
+    body: parameters,
+  });
+
+  try {
+    assert(response, TokenResponseSchema);
+    const { access_token, refresh_token, id_token, scope, expires_in } = response;
```
Comment (simeng-li) [inline packages/js/src/core/fetch-token.ts:60]: nit: `{ access_token: accessToken, refresh_token: refreshToken ... } = response; `

File: packages/js/src/core/fetch-token.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  refreshToken: string;
+  resource?: string;
+  scope?: string[];
+}
+
+const TokenResponseSchema = type({
+  access_token: string(),
+  refresh_token: string(),
+  id_token: string(),
+  scope: string(),
```
Comment (simeng-li) [inline packages/js/src/core/fetch-token.ts:20]: I am not quite sure whether the scope return value is required or not. If no scope if provided, I believe it could be undefined. Let's go with the optional for now. I'll update the doc.

File: packages/playground/package.json (hunk 5 above)
Comment (demonzoo) [reply packages/playground/package.json:13]: No idea. I would also like to know why. 😂
Anyways, I will try to remove this change from this PR, and open up a new one.

File: packages/js/src/core/fetch-token.ts (hunk 8 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:20]: ok, makes sense to me

File: packages/js/src/core/fetch-token.ts (hunk 6 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:53]: I agree. Should I put it in a `js/consts/index.ts` file?

File: packages/js/jest.config.js (hunk 1 above)
Comment (wangsijie) [reply packages/js/jest.config.js:6]: js core SDK can be run in node environment.

File: packages/js/src/core/fetch-token.ts (hunk 2 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:41]: Done.

File: packages/js/src/core/fetch-token.ts (hunk 7 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts:60]: I tried, but personally don't quite like the outcome. For me it does NOT seem to increase readability.

image

File: packages/js/src/core/fetch-token.ts (hunk 3 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.ts]: Uh-huh, I see your points. Code updated to allow user to pass in a fetchFunction. Please let me know if this solves the problem. Thx.

File: packages/playground/package.json (hunk 5 above)
Comment (demonzoo) [reply packages/playground/package.json]: Interesting enough, I reverted the changes made to disable the interactive build, but this time the CI is working...
Not sure what went wrong earlier, but now it seems the changes will no longer be needed, at least for now

Comment (demonzoo) [reply packages/playground/package.json]: @gao-sun you can resolve this comment now. The code change has been reverted.

**Review by simeng-li — APPROVED** [review]
LGTM

File: packages/js/src/consts/index.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export const CONTENT_TYPE = {
+  FORM_URL_ENCODED: { 'Content-Type': 'application/x-www-form-urlencoded' },
```
Comment (gao-sun) [inline packages/js/src/consts/index.ts:2]: we don't have a convention of using upper snake case for constants. Just use camelCase for the key and PascalCase for the object name is fine.

File: packages/js/src/core/fetch-token.test.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  test('get token response', async () => {
+    const mockedOidcResponse = {
+      access_token: 'access_token',
+      refresh_token: 'refresh_token',
+      id_token: 'id_token',
+      scope: 'read register manage',
+      expires_in: 3600,
+    };
+
+    const expectedTokenResponse: RefreshTokenTokenResponse = {
```
Comment (gao-sun) [inline packages/js/src/core/fetch-token.test.ts:15]: use camelcaseKeys？

**Review by gao-sun — APPROVED** [review]
overall looks good, please resolve conversation before merging

File: packages/js/jest.config.js (hunk 1 above)
Comment (demonzoo) [reply packages/js/jest.config.js:6]: After offline team discussion, we decided to run tests for both jsdom and node envs, by executing the `run test` command twice in parallel in CI. Thus remove this env setting from jest.config.js

File: packages/js/src/core/fetch-token.test.ts (hunk 10 above)
Comment (demonzoo) [reply packages/js/src/core/fetch-token.test.ts:15]: `camelcaseKeys` or `camelCaseKeys`?

File: packages/js/src/consts/index.ts (hunk 9 above)
Comment (demonzoo) [reply packages/js/src/consts/index.ts]: done
